v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/bigtablechangestreamstobigquery/BigtableChangeStreamsToBigQuery.java

```diff
@@ -0,0 +1,436 @@
+  }
+
+  private static String getBigtableProjectId(BigtableChangeStreamsToBigQueryOptions options) {
+    return StringUtils.isEmpty(options.getBigtableProjectId())
+        ? options.getProject()
+        : options.getBigtableProjectId();
+  }
+
+  private static String getBigQueryChangelogTableName(
+      BigtableChangeStreamsToBigQueryOptions options) {
+    return StringUtils.isEmpty(options.getBigQueryChangelogTableName())
+        ? options.getBigtableTableId() + "_changelog"
+        : options.getBigQueryChangelogTableName();
+  }
+
+  private static String getBigQueryProjectId(BigtableChangeStreamsToBigQueryOptions options) {
+    return StringUtils.isEmpty(options.getBigQueryProjectId())
+        ? options.getProject()
+        : options.getBigQueryProjectId();
+  }
+
+  /**
+   * Remove the following intermediate metadata fields that are not user data from {@link TableRow}:
+   * _metadata_error, _metadata_retry_count, _metadata_original_payload_json.
+   */
+  private static TableRow removeIntermediateMetadataFields(TableRow tableRow) {
+    TableRow cleanTableRow = tableRow.clone();
+    Set<String> allColumns = tableRow.keySet();
+
+    for (String column : allColumns) {
+      if (TransientColumn.isTransientColumn(column)) {
+        cleanTableRow.remove(column);
+      }
+    }
+
+    return cleanTableRow;
+  }
+
+  /**
+   * DoFn that converts a {@link ChangeStreamMutation} to multiple {@link Mod} in serialized JSON
+   * format.
+   */
+  static class ChangeStreamMutationToModJsonFn extends DoFn<ChangeStreamMutation, String> {
+
+    private final BigtableSource sourceInfo;
+
+    ChangeStreamMutationToModJsonFn(BigtableSource source) {
+      this.sourceInfo = source;
+    }
+
+    @ProcessElement
+    public void process(@Element ChangeStreamMutation input, OutputReceiver<String> receiver)
+        throws Exception {
+      for (Entry entry : input.getEntries()) {
+        ModType modType = getModType(entry);
+
+        Mod mod = null;
+        switch (modType) {
+          case SET_CELL:
+            mod = new Mod(sourceInfo, input, (SetCell) entry);
+            break;
+          case DELETE_CELLS:
+            mod = new Mod(sourceInfo, input, (DeleteCells) entry);
+            break;
+          case DELETE_FAMILY:
+            mod = new Mod(sourceInfo, input, (DeleteFamily) entry);
+            break;
+          default:
+          case UNKNOWN:
+            throw new UnsupportedEntryException("Cloud Bigtable change stream entry of type " +
+                entry.getClass().getName()
+                + " is not supported. The entry was put into a dead letter queue directory. "
+                + "Please update your Dataflow template with the latest template version");
+        }
+
+        String modJsonString;
+
+        try {
+          modJsonString = mod.toJson();
+        } catch (IOException e) {
+          // Ignore exception and print bad format.
+          modJsonString = String.format("\"%s\"", input);
```

Comment:
is it okay that this won't be valid json? also should we be logging some error here?

Reply:
This should not happen and this code is the same as in Spanner's ChangeStreams -> BQ. What will happen is that this Mod will be written as a JSON into a DLQ directory so we will be able to see what went down